pageserver/src/basebackup.rs

```diff
@@ -4,181 +4,261 @@
+            // User defined tablespaces are not supported
+            assert!(db.spcnode == pg_constants::DEFAULTTABLESPACE_OID);
+            let src_path = format!("{}/base/1/PG_VERSION", self.snappath);
+            let dst_path = format!("base/{}/PG_VERSION", db.dbnode);
+            self.ar.append_path_with_name(&src_path, &dst_path)?;
+            format!("base/{}/pg_filenode.map", db.dbnode)
+        };
+        assert!(img.len() == 512);
+        let header = new_tar_header(&path, img.len() as u64)?;
+        self.ar.append(&header, &img[..])?;
+        Ok(())
     }
 
-    // FIXME: Also send all the WAL. The compute node would only need
-    // the WAL that applies to non-relation files, because the page
-    // server handles all the relation files. But we don't have a
-    // mechanism for separating relation and non-relation WAL at the
-    // moment.
-    for entry in std::fs::read_dir(&walpath)? {
-        let entry = entry?;
-        let fullpath = &entry.path();
-        let relpath = fullpath.strip_prefix(&walpath).unwrap();
-
-        if !entry.path().is_file() {
-            continue;
+    //
+    // Extract twophase state files
+    //
+    fn add_twophase_file(&mut self, tag: &ObjectTag, xid: TransactionId) -> anyhow::Result<()> {
+        // Include in tarball two-phase files only of in-progress transactions
+        if self.timeline.get_tx_status(xid, self.lsn)?
+            == pg_constants::TRANSACTION_STATUS_IN_PROGRESS
+        {
+            let img = self.timeline.get_page_at_lsn_nowait(*tag, self.lsn)?;
+            let mut buf = BytesMut::new();
+            buf.extend_from_slice(&img[..]);
+            let crc = crc32c::crc32c(&img[..]);
+            buf.put_u32_le(crc);
+            let path = format!("pg_twophase/{:>08X}", xid);
+            let header = new_tar_header(&path, buf.len() as u64)?;
+            self.ar.append(&header, &buf[..])?;
         }
-
-        let archive_fname = relpath.to_str().unwrap();
-        let archive_fname = archive_fname
-            .strip_suffix(".partial")
-            .unwrap_or(&archive_fname);
-        let archive_path = "pg_wal/".to_owned() + archive_fname;
-        ar.append_path_with_name(fullpath, archive_path)?;
+        Ok(())
     }
 
-    ar.finish()?;
-    debug!("all tarred up!");
-    Ok(())
+    //
+    // Add generated pg_control file
+    //
+    fn add_pgcontrol_file(&mut self) -> anyhow::Result<()> {
+        let checkpoint_bytes = self
+            .timeline
+            .get_page_at_lsn_nowait(ObjectTag::Checkpoint, self.lsn)?;
+        let pg_control_bytes = self
+            .timeline
+            .get_page_at_lsn_nowait(ObjectTag::ControlFile, self.lsn)?;
+        let mut pg_control = ControlFileData::decode(&pg_control_bytes)?;
+        let mut checkpoint = CheckPoint::decode(&checkpoint_bytes)?;
+
+        // Generate new pg_control and WAL needed for bootstrap
+        let checkpoint_segno = self.lsn.segment_number(pg_constants::WAL_SEGMENT_SIZE);
```

Comment:
So we're still generating a WAL segment out of thin air, and filling it with data - the checkpoint record - in the page server. This has all the same problems as the earlier pg_resetwal approach. It's not clear how the wal safekeeper or replicas should react if there is a gap in the WAL sequence, or if some piece of WAL is created twice with different contents.

I thought we had a solution that doesn't require this, with the `zenith.signal` file, that doesn't require modifying the WAL. Let's write the checkpoint struct directly in the zenith.signal file, and read it in Postgres startup from there.

Reply:
New approach only seems to be very similar with old approach leaving holes in WAL.
But actually we are not changing current LSN.
Yes, we generate WAL segment with single XLOG_CHECKPOINT_SHUTDOWN record.
But checkpointCopy.redo saved in this record point to the current LSN.
So after fetching this checkpoint Postgres continue to append WAL from this LSN.

My original intention was to eliminate reading checkpoint from WAL during startup at all.
But then I found out that it requires too much changes in Postgres core because while fetching checkpoint Postgres also initialize WAL reader which is needed for further work. So I decided to generate WAL segment and let Postgres extract checkpoint record from it in standard ways. Unfortunately still I need to make some changes in core code, to let Postgres accept situation when address of checkpint record and redo point are different.

So briefly speaking, current approach doesn't require changing of WAL and can be used both at master and replicas.
The only concern that the beginning of WLA segment contains "garbage". But if we enforce that wal proposer start replication only from current LSN (checkpointCopy.redo) and not trying to send data before this point, then there isno risk that walkeepers will receive dummy data which may overwrite valid WAL records.

Reply:
To initialize the xlogreader, you just need to call `XLogBeginRead(xlogreader, checkpointCopy.redo)`, no? 

Hmm, is `xl_prev` set correctly for the first record that's created after startup?

Reply:
> To initialize the xlogreader, you just need to call XLogBeginRead(xlogreader, checkpointCopy.redo), no?
Not only. I do not remember of details now. For example I got assertion failure because expectedTLEs was  not initialized.
Certainly it is possible to let postgres proceed startup without actual reading WAL. May be it will require more changes in core, may be not. I didn't consider it as principle moment. For me the most important thing was to start server without leaving gap in WAL. If we really want to avoid generation of any WAL, it can be done.

Reply:
> Hmm, is xl_prev set correctly for the first record that's created after startup?

Yes, it is not correctly set now. Thinking how to handle it...

Reply:
I committed solution fixing xl_prev problem. I just added tracking of prev record LSN position to page server (in the same way as it is handling last_record_lsn now). I am not 100% that both of them can survive page server crash, because this fields are persistent in timeline metadata only by checkpoint or at branch creation time. But at least the guarantees for  last_record_lsn are the same.

Reply:
Hmm, that only works if you create a branch at the end of another branch. If you try to create a branch at an older point in time, you would need to know the beginning of the previous record at the branch point. The 'test_branch_behind' test tests that case.

The case where incorrect xl_prev would cause trouble is when you would have a standby server following the branch across the record with incorrect xl_prev. Or any other tool that reads the WAL stream. `pg_waldump` would complain, I presume. Please consider how to test this.

Reply:
Yes, you are right. I have added assert check in previous commit and have to immediately removed it because test_branch_behind is not passed. Two comments:

1. xl_prev is not checkout now by our wal reader, so incorrect value of this file doesn't prevent page server from accepting this WAL. Yes, pg_waldump will not work and most likely replicas will also refuse records with incorrect xl_prev
2. The problem is actually caused by the way of creating branches now: we just point arbitrary LSN and which branching should be done. If instead of this model we allow users to create "snapshots" (recovery points), then we can safely remember prev_record_ptr at the moment of snapshot creation.

I do not know any other solution of the problem with existed branching model rather than keep or fetch requested WAL segment by page server and include it in tarball sent to started computation node instead of generated WAL segment.
As far as I understand we want to avoid maintaining WAL at pageserver is much as possible.
This is why I tried to provide solution which doesn't require to extract correspondent WSL segment.

Reply:
> The problem is actually caused by the way of creating branches now: we just point arbitrary LSN and which branching should be done.

Good point. Currently, you can start branch in the middle of a WAL record. There's no way to construct a valid WAL segment with that.

> If instead of this model we allow users to create "snapshots" (recovery points), then we can safely remember prev_record_ptr at the moment of snapshot creation.

Hmm, that's not good enough for point in time recovery. We could store the LSNs of every commit record, and only create branches at commit records, though. So every commit record would implicitly be a snapshot or recovery. That would require a fair amount of new code...

> I do not know any other solution of the problem with existed branching model rather than keep or fetch requested WAL segment by page server and include it in tarball sent to started computation node instead of generated WAL segment.
As far as I understand we want to avoid maintaining WAL at pageserver is much as possible.
This is why I tried to provide solution which doesn't require to extract correspondent WSL segment.

I think it would be OK to keep the *last* WAL segment on each branch in the page server. Or fetch it from the WAL safekeeper when you restart Postgres. But keeping *all* WAL, just in case you want to create a branch later, seems unreasonable.

I think we could have different mechanisms for restarting a compute node at the tip of a branch, and for creating a new branch. When you create/restart a compute node at the tip of the branch, tracking the last record LSN like you did here works. And creating a new branch at the end of an existing branch could do the same thing. But when you create a branch at an earlier point in time, set xl_prev to 0, and start with a new empty `WAL` segment, as if you ran `pg_resetwal`. Because of that point of "discontinuity" or "gap" in the WAL, you cannot have a standby server that follows the switch from the parent branch to the new branch, as it would trip because of the gap in the WAL. But I think that's OK.